### ISSUE TITLE
Unable to read lat/lon from ECMWF ERA5 grib file

### ISSUE THREAD

**Issue by blair-anson** [Unidata/netcdf-java#297]
To report a non-security related issue, please provide:

* the version of the software with which you are encountering an issue
5.3.2

* environmental information (i.e. Operating System, compiler info, java version, python version, etc.)
Ubuntu
java8

* a description of the issue with the steps needed to reproduce it
- download an ERA5 grib file from ECMWF  (I have attached an example)
https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-single-levels?tab=form
- open it using the following code

```
NetcdfFile ncfile = NetcdfFile.open(filename);
Variable lon = ncfile.findVariable("lon");
Variable lat = ncfile.findVariable("lat");
Array data_lat = lat.read();
Array data_lon = lon.read();
```
data_lon and data_lat should be an Array containing the latitude & longitude values but instead they contain a "NaN" value.

By comparison, if I use the wgrib command line tool (apt-get install grads) I can see the first record is as below, where the lat/lon is correctly shown.....

```
wgrib example_197901.grib -V | head
rec 1:0:date 1979010100 FLSR kpds5=245 kpds6=1 kpds7=0 levels=(0,0) grid=255 sfc ECext=1 class=23 type=analysis stream=1025 Version=0001 anl:
  FLSR=Forecast logarithm of surface roughness for heat []
  timerange 0 P1 0 P2 0 TimeU 1  nx 1 ny 1 GDS grid 0 num_in_ave 0 missing 0
  center 98 subcenter 0 process 145 Table 128 scan: WE:NS winds(N/S) 
  latlon: lat  18.950000 to 18.950000 by 0.250000  nxny 1
          long 76.210000 to 76.210000 by 0.250000, (1 x 1) scan 0 mode 128 bdsgrid 1
  min/max data -5.42122 -5.42122  num bits 16  BDS_Ref -5.42122  DecScale 0 BinScale 0
```

[example_197901.zip](https://github.com/Unidata/netcdf-java/files/4682718/example_197901.zip)

**Comment by moisespr123** [comment]
The CDMWF files are somewhat encoded differently. They recommend using ecCode to read them. Alternatively, you can download the ECMWF files straight in NetCDF. Even although they state NetCDF support is experimental, I haven't had any issues sending requests. Therefore reducing the hassle of working with the Grib files.

**Comment by lesserwhirls** [comment]
Netcdf-java is having issues with the fact that the grib1 messages have encoded the number of points in the latitude and longitude directions to each be 1 (as seen by the `nx` and `ny` parameters extracted by wgrib), and so the grid is being defined as having a shape of 1 x 1. That is causing issues in other parts of the code, like constructing the latitude and longitude variables.  While you can see the bounds of the grid in the wgrib output (and netcdf-java is able to read those as well), it too indicates that the grid is 1x1. I might be able to extend our code to deal with this particular situation, but it'd be better for everyone involved if the messages were encoded properly.

**Comment by blair-anson** [comment]
@lesserwhirls There is only 1 lat/lng point in the grib file I provided, so the grib message is correct. I frequently need to deal with 1 lat/lng point of data, and less often a large grid of data. Downloading from ECMWF is quite slow, so I try to save time by downloading only 1 point when doing initial investigations of a various locations, and then I expand to larger areas if the initial investigation is fruitful. 
If it is possible to handle this edge case of 1 point, then that would be fantastic!

Possible related issue; I also had a problem with NetCDF files that have 1 lat/lng point (see attached example netcdf file). 
Calling  getRangesFromLatLonRect() throws a... 
```
ucar.ma2.InvalidRangeException: Request Bounding box does not intersect Grid 
at ucar.nc2.dt.grid.GridCoordSys.getRangesFromLatLonRect(GridCoordSys.java:1153)
```

Example code...

```
public void read_range() throws IOException, InvalidRangeException {
        // same bounding box that was used to request the data from ECMWF
        BBox bbox = BBox.builder()
                .bottomRightLat(18.94)
                .bottomRightLng(76.23)
                .topLeftLat(18.97)
                .topLeftLng(76.2)
                .build();

        String filename = "./example_197901_netcdf.nc";

        List<Range> geo_index_range = null;
        try (GridDataset gds = GridDataset.open(filename)) {
            GridDatatype grid = gds.findGridDatatype("ssrd");  // example variable
            GridCoordSystem gcs = grid.getCoordinateSystem();

            LatLonRect geo_area = new LatLonRect(
                    new LatLonPointImpl(
                            bbox.getTopLeftLat(),
                            bbox.getTopLeftLng()),
                    new LatLonPointImpl(
                            bbox.getBottomRightLat(),
                            bbox.getBottomRightLng())
            );
            System.out.println(String.format("Lat max %s  Lat min %s, Lon max %s, Lon min %s",
                    geo_area.getLatMax(),
                    geo_area.getLatMin(),
                    geo_area.getLonMax(),
                    geo_area.getLonMin()
            ));

            geo_index_range = gcs.getRangesFromLatLonRect(geo_area);  // error here
        }

        System.out.println(geo_index_range.size());
    }
```
[example_197901_netcdf.zip](https://github.com/Unidata/netcdf-java/files/4686457/example_197901_netcdf.zip)

**Comment by blair-anson** [comment]
@moisespr123 yes I have tried their netcdf files too. I found them to be slower to download than the grib files, so although I like dealing with netcdf better than grib the improvement in download time is worth the hassle

**Comment by lesserwhirls** [comment]
Hmmm. I don't think I've encountered a GRIB message with a single point before. At this point I have a fix that will let you interact with these GRIB messages as a `NetcdfFile` or `NetcdfDataset`. Given that the data are essentially a timeseries from a single grid point, reading the file as if it were a Gridded dataset (i.e. using `GridDataset`) will not work.

**Comment by blair-anson** [comment]
@lesserwhirls Amazing thank you! I only use `GridDataset` to identify what lat lon ranges are needed. If I first use `NetcdfFile `or `NetcdfDataset` to check if the length of the lat / lon variables equal 1, then I will skip using `GridDataset` and just focus on calculating the time range I need.

A related question: what is the recommended way for changing the longitude range from 0:360 to -180:180?  I currently use `LatLonPointImpl.lonNormal()` on each lon value, but is there a better approach?

**Comment by cofinoa** [comment]
@blair-anson as @lesserwhirls mentions the GRIB IOSP assumes a GRID on the GRIB message. 

You should request at least 2 by 2 grid points. The time on requesting that to the CDS will be the same as 1 by 1 case, and the size increase will be marginal (+7.5% with respect to the 1 by 1).

**Comment by blair-anson** [comment]
@cofinoa yes that workaround would be fine too

**Comment by lesserwhirls** [comment]
> A related question: what is the recommended way for changing the longitude range from 0:360 to -180:180? I currently use LatLonPointImpl.lonNormal() on each lon value, but is there a better approach?

There isn't much to the method `LatLonPointImpl.lonNormal()`, and it relies on the assumption that it's looking at a longitude value with units of `degrees_east`. Here is what it is doing:

~~~java
public static double lonNormal(double lon) {
  if ((lon < -180.) || (lon > 180.)) {
    return Math.IEEEremainder(lon, 360.0);
  } else {
    return lon;
  }
}
~~~

**Comment by dopplershift** [comment]
Fixed by #302?

**Comment by JohnLCaron** [comment]
Can we close this @lesserwhirls?
